exec/cnex/httpserver.c

```diff
@@ -0,0 +1,353 @@
+                break;
+            }
+            string_freeString(value);
+        }
+    } else {
+        goto cleanup;
+    }
+    client_handleResponse(c, response);
+    retval = TRUE;
+
+cleanup:
+    string_freeStringArray(lines);
+    string_freeStringArray(req);
+    response_freeResponse(response);
+    return retval;
+}
+
+void client_handleResponse(struct tagTClient *c, struct tagTHttpResponse *response)
+{
+    if (response->code == 0) {
+        // If the code is 0, that means there is no body, nothing to return; so set the code to 204 (No Content)
+        // and respond with it.
+        response->code = 204;
+    }
+    TString *r = response_toString(response);
+    send(c->socket, r->data, (int)r->length, 0);
+    string_freeString(r);
+}
+
+BOOL client_headerMatch(TString *header, const char *target, TString **value)
+{
+    char *t = strdup(target);
+    size_t target_len = strlen(t);
+
+    TString *h = string_subString(header, 0, target_len);
+
+    // The following will convert the strings to lowercase in place, rather than creating a whole new string
+    // in the process like string_toLowerCase() does.  (There's no reason to do that.)
+    for (size_t c = 0; c < h->length; c++) {
+        h->data[c] = (char)tolower(h->data[c]);
+    }
+    for (size_t c = 0; c < target_len; c++) {
+        t[c] = (char)tolower(t[c]);
+    }
+    if (string_compareCString(h, t) == 0) {
+        free(t);
+        size_t i = target_len;
+        size_t c = 0;
+        while (i < header->length && isspace(header->data[i])) {
+            ++i;
+            ++c;
+        }
+        *value = string_subString(header, i, header->length - i);
+        string_freeString(h);
+        return TRUE;
+    }
+    string_freeString(h);
+    free(t);
+    return FALSE;
+}
+
+void client_freeClient(struct tagTClient *c)
+{
+    closesocket(c->socket);
+    string_freeString(c->request);
+    string_freeString(c->post_data);
+    string_freeString(c->path);
+    free(c);
+    c = NULL;
+}
+
+
+
+HttpServerImpl *server_createServer(unsigned short port, IHttpServerHandler *handler)
+{
+    HttpServerImpl *server = malloc(sizeof(HttpServerImpl));
+    server->handler = handler;
+    server->clients = NULL;
+    server->num_clients = 0;
+
+#ifdef _WIN32
+    static BOOL initialized = FALSE;
+    if (!initialized) {
+        WSADATA wsa;
+        WSAStartup(MAKEWORD(1, 1), &wsa);
+        initialized = TRUE;
+    }
+#endif
+    server->server = socket(PF_INET, SOCK_STREAM, 0);
+    assert(server->server != INVALID_SOCKET);
+    int on = 0;
+    int r = setsockopt(server->server, SOL_SOCKET, SO_REUSEADDR, (char *)&on, sizeof(on));
+    assert(r == 0);
+    struct sockaddr_in sin;
+    sin.sin_family = AF_INET;
+    sin.sin_addr.s_addr = INADDR_ANY;
+    sin.sin_port = htons(port);
+    r = bind(server->server, (struct sockaddr *)&sin, sizeof(sin));
+    if (r != 0) {
+        // The executor will continue to run; just without debugger support.
```

Comment:
This comment doesn't seem to be true, because if this function returns false the executor mostly continues on as if debugging support actually is available.

Reply:
Changed to a `fatal_error()` to avoid the debugger starting, but never actually running any code.